File: data/r2dbc/r2dbc-flyway-postgres/src/main/java/com/mahmud/r2dbcflywaypostgres/controller/TodoController.java
package com.mahmud.r2dbcflywaypostgres.controller;

import com.mahmud.r2dbcflywaypostgres.domain.Todo;
import com.mahmud.r2dbcflywaypostgres.service.TodoService;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@RestController
@RequestMapping("/api/todos")
public class TodoController {

    private final TodoService todoService;

    public TodoController(TodoService todoService) {
        this.todoService = todoService;
    }

    @GetMapping
    public Flux<Todo> getAllTodos() {
        return todoService.getAllTodos();
    }

    @GetMapping("/completed/{completed}")
    public Flux<Todo> getTodosByCompleted(@PathVariable boolean completed) {
        return todoService.getTodosByCompleted(completed);
    }

    @GetMapping("/{id}")
    public Mono<Todo> getTodoById(@PathVariable Long id) {
        return todoService.getTodoById(id);
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public Mono<Todo> createTodo(@RequestBody Todo todo) {
        return todoService.createTodo(todo);
    }

    @PutMapping("/{id}")
    public Mono<Todo> updateTodo(@PathVariable Long id, @RequestBody Todo todo) {
        return todoService.updateTodo(id, todo);
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public Mono<Void> deleteTodo(@PathVariable Long id) {
        return todoService.deleteTodo(id);
    }
}

File: io/grpc-mongodb-todo/grpc-client/src/main/kotlin/com/mahmud/grpcclient/controller/TodoController.kt
package com.mahmud.grpcclient.controller


import com.mahmud.grpcclient.Empty
import com.mahmud.grpcclient.service.TodoClientService
import com.mahmud.grpcclient.TodoItem
import com.mahmud.grpcclient.dto.TodoItemDto
import com.mahmud.grpcclient.dto.TodoRequestDto
import com.mahmud.grpcclient.dto.mapToTodoItemDto
import org.springframework.web.bind.annotation.*
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.time.Instant

@RestController
@RequestMapping("/api/todos")
class TodoController(private val todoService: TodoClientService) {

    @PostMapping
    fun createTodo(
        @RequestBody request: TodoRequestDto
    ): Mono<TodoItemDto> {
        val todo = todoService.createTodo(
            request.title,
            request.description,
            request.completed,
            request.dueDate?.let { Instant.ofEpochSecond(it) }
        )
         return Mono.justOrEmpty(todo?.let { mapToTodoItemDto(it) })
    }

    @GetMapping("/{id}")
    fun getTodo(@PathVariable id: String): Mono<TodoItemDto> {
        return Mono.justOrEmpty(mapToTodoItemDto(todoService.getTodo(id)))
    }

    @GetMapping
    fun listTodos(@RequestParam(defaultValue = "true") showCompleted: Boolean): Flux<TodoItemDto> {
        val todoList = todoService.listTodos(showCompleted)
        return todoList?.let { Flux.fromIterable(it).map { todoItem -> mapToTodoItemDto(todoItem) } } ?: Flux.empty()
    }

    @PutMapping("/{id}")
    fun updateTodo(
        @PathVariable id: String,
        @RequestBody request: TodoRequestDto
    ): Mono<TodoItemDto> {
        val todo = todoService.updateTodo(
            id,
            request.title,
            request.description,
            request.completed,
            request.dueDate?.let { Instant.ofEpochSecond(it) }
        )
        return Mono.justOrEmpty(todo?.let { mapToTodoItemDto(it) })
    }

    @DeleteMapping("/{id}")
    fun deleteTodo(@PathVariable id: String): Empty? {
        return todoService.deleteTodo(id)
    }
}